### DISCUSSION TITLE
Combat, functioning of the ghost ability (version 1.0.1 6629)

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#6866] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Ghosts that are incorporated into a ghost group through the ghost ability should not replace the ghosts that have been decimated but should be added to the total of the original ghost group.

**First example (video below).**
The first group of ghosts consists of twenty five ghosts.
Five ghosts are decimated by the cold ray spell.
Twenty ghosts remain in the group.
The group of ghosts attacks the crusaders.
One crusader is decimated.
The group of ghosts grows to twenty-one ghosts.
The group of crusaders retaliates.
Seven ghosts are decimated.
Fourteen ghosts remain in the group.
The animate dead spell reanimates the ghosts for a total group of twenty five ghosts.
However, the group should have twenty-six ghosts, because a crusader has been transformed into a ghost.

https://user-images.githubusercontent.com/43583994/223468064-2afcc221-fee1-4cf8-b57b-045e9ef670ae.mp4

**Second example (video below).**
The first group of ghosts consists of twenty five ghosts.
The group of ghosts attacks the crusaders.
One crusader is decimated.
The group of ghosts grows to twenty six ghosts.
The group of crusaders retaliates.
Ten ghosts are decimated.
Sixteen ghosts remain in the group.
The animate dead spell reanimates the ghosts for a total group of twenty six ghosts.
In this case, this has been taken into account.

https://user-images.githubusercontent.com/43583994/223469348-dd507d3a-6b96-4006-82e9-23344c7ab705.mp4

**Third example (video below).**
The first group of ghosts consists of twenty five ghosts.
Five ghosts are decimated by the cold ray spell.
There are twenty ghosts left in the group.
The animate dead spell could be used on this group of ghosts to reanimate the five ghosts lost before.
The group of ghosts attacks the peasants.
Twenty five peasants are decimated.
The group of ghosts grows to forty five ghosts.
However, the animate dead spell cannot be used on this group of ghosts to reanimate the five ghosts lost before.

https://user-images.githubusercontent.com/43583994/223470023-172c6e80-a7bc-4e47-a506-faee96cf716d.mp4

**Fourth example (video below).**
The first group of ghosts consists of twenty five ghosts.
The group of ghosts attacks the peasants.
Twenty-five peasants are decimated.
The group of ghosts grows to fifty ghosts.
The group of crusaders attacks the group of ghosts successively until the group of crusaders is decimated by the ghosts group's retaliation.
There are then nineteen ghosts left in the group.
The animate dead spell reanimates the ghosts for a total group of fifty ghosts.
However, the group should have fifty-five ghosts, because five crusaders have been transformed into ghosts.

https://user-images.githubusercontent.com/43583994/223470616-3e1ed01f-c5a9-4ffb-bb22-2fe9c2c89d63.mp4

**There is a bug in the original game (video below).**

https://user-images.githubusercontent.com/43583994/223471622-e7eef228-ef74-4670-9370-9a98b96e59af.mp4

### Save file

Here are the scenario test and save files:
[Functioning of the ghost ability.zip](https://github.com/ihhub/fheroes2/files/10911316/Functioning.of.the.ghost.ability.zip)

**Comment by oleg-derevenetz** [comment]
> The first group of ghosts consists of twenty five ghosts.
> Five ghosts are decimated by the cold ray spell.
> Twenty ghosts remain in the group.
> The group of ghosts attacks the crusaders.
> One crusader is decimated.
> The group of ghosts grows to twenty-one ghosts.
> The group of crusaders retaliates.
> Seven ghosts are decimated.
> Fourteen ghosts remain in the group.
> The animate dead spell reanimates the ghosts for a total group of twenty five ghosts.
> However, the group should have twenty-six ghosts, because a crusader has been transformed into a ghost.

That's not how it works in the original game:

https://user-images.githubusercontent.com/32623900/223495275-987ef384-f691-488f-894d-10b35c20f6fe.mp4

What happens here is that some decimated ghost is literally replaced by a ghost made from Crusader, and no ghosts are added above the initial number.

> There is a bug in the original game (video below).

I'm not sure that this is a bug, most likely it is not. There is a certain logic that is reproduced in fheroes2. Perhaps the logic in the original game is made this way precisely so that it would be impossible to abuse Animate Dead with ghosts.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> > The first group of ghosts consists of twenty five ghosts.
> > Five ghosts are decimated by the cold ray spell.
> > Twenty ghosts remain in the group.
> > The group of ghosts attacks the crusaders.
> > One crusader is decimated.
> > The group of ghosts grows to twenty-one ghosts.
> > The group of crusaders retaliates.
> > Seven ghosts are decimated.
> > Fourteen ghosts remain in the group.
> > The animate dead spell reanimates the ghosts for a total group of twenty five ghosts.
> > However, the group should have twenty-six ghosts, because a crusader has been transformed into a ghost.
> 
> That's not how it works in the original game:
> 
> What happens here is that some decimated ghost is literally replaced by a ghost made from Crusader, and no ghosts are added above the initial number.
> 
> > There is a bug in the original game (video below).
> 
> I'm not sure that this is a bug, most likely it is not. There is a certain logic that is reproduced in fheroes2. Perhaps the logic in the original game is made this way precisely so that it would be impossible to abuse Animate Dead with ghosts.

I consider this a bug.
Since what is the point of allowing the hero to cast a spell (in this case animate dead) that will not have the intended effect of the spell (in the original game example that removes ghosts from the group that attacked the peasants)?
In this case, the spell should not be able to be cast at all, with the message window displaying; "That spell will affect no one!" (as the number of ghosts in the group is at least as great as the number of ghosts at the start of the combat).

**Comment by oleg-derevenetz** [comment]
> I consider this a bug.

This particular case in last video is indeed a bug, and it is fixed in fheroes2:

https://user-images.githubusercontent.com/32623900/223514912-2b027b59-e656-4fc3-a6d9-cca9cc2d7dd2.mp4

But the OG behavior when new ghosts "replace" the fallen ones instead of "adding" on top of the original number is not a bug, this is for reason - most likely to not allow heroes with Ghosts in their armies to abuse the Animate Dead too much.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> > I consider this a bug.
> 
> This particular case in last video is indeed a bug, and it is fixed in fheroes2:
> 
> But the OG behavior when new ghosts "replace" the fallen ones instead of "adding" on top of the original number is not a bug, this is for reason - most likely to not allow heroes with Ghosts in their armies to abuse the Animate Dead too much.

The ghost's ability is that it transforms decimated creatures into ghosts, it is not that it regenerates like the vampire lord would.

**Comment by oleg-derevenetz** [comment]
> The ghost's ability is that it transforms decimated creatures into ghosts, it is not that it regenerates like the vampire lord would.

But that's how it works in the original game. Even if you don't like this, this is not necessary a bug :) This is just a difference of opinion between you and the developers of the original game.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> > The ghost's ability is that it transforms decimated creatures into ghosts, it is not that it regenerates like the vampire lord would.
> 
> But that's how it works in the original game. Even if you don't like this, this is not necessary a bug :) This is just a difference of opinion between you and the developers of the original game.

The possibility for a player to have ghosts in the army of one of his heroes appeared with the expansion "The Price of Loyalty".
Everything was not necessarily perfect considering all the bugs still present in the game.
We will certainly never know how the devolopers would like to see the ghosts work in this regard. :)

But if we stick to what we know, "Creatures killed by ghosts become ghosts", it is nowhere written (to my knowledge) that ghosts have the same ability to regenerate as vampire lords.
It would be nice to have a clear functioning on this subject.

Edit:
**Fifth example (video below).**
The first group of ghosts is composed of twenty-five ghosts.
The group of ghosts attacks the central group of five pikemen.
Five pikemen are decimated.
The group of ghosts grows to thirty ghosts.
The group of five pikemen at the top of the screen attacks the ghosts, without success.
The group of ghosts retaliates and decimates five pikemen.
The group of ghosts grows to thirty-five ghosts.
The other three groups of pikemen attack the group of ghosts in succession.
Twenty-two ghosts remain in the group.
The animate the dead spell reanimates thirteen ghosts for a total group of thirty five ghosts.
Again, the incorporated ghosts have been added to the original group of ghosts, as is also the case in example two above.

https://user-images.githubusercontent.com/43583994/223605312-f16f66f7-e24c-4af6-a55a-99d4c1e42fc7.mp4

Anyway, it doesn't make much sense for the ghost to have both abilities, especially since the "Soul Eater" ability is much better than the "Drain of Life" ability.
And, in addition, the "Drain of Life" ability reduces the effectiveness of the "Soul Eater" ability by preventing the reanimation of incorporated ghosts.

**Comment by oleg-derevenetz** [comment]
> And, in addition, the "Drain of Life" ability reduces the effectiveness of the "Soul Eater" ability by preventing the reanimation of incorporated ghosts.

Yes, of course. Without this reduction ghosts would be even bigger imba for a hero with the Animate Dead spell. Still don't see any bug here. The only bug I see is that in the original game the Animate Dead spell reduces the number of ghosts if there are already extra ghosts in a stack, and this bug is already fixed in fheroes2.

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> > And, in addition, the "Drain of Life" ability reduces the effectiveness of the "Soul Eater" ability by preventing the reanimation of incorporated ghosts.
> 
> Yes, of course. Without this reduction ghosts would be even bigger imba for a hero with the Animate Dead spell. Still don't see any bug here. The only bug I see is that in the original game the Animate Dead spell reduces the number of ghosts if there are already extra ghosts in a stack, and this bug is already fixed in fheroes2.

There will be other ways to handle ghosts than having a ghost functioning that includes the vampire lord ability mixed with the ghost ability.

It is not clear to the player:
"When can I use the Animate Dead spell?"
"Before attacking other creatures?"
"Will the spell work after the attack?"

It's best to start with a simple and solid foundation:
Ghost ability is "Soul Eater", "Creatures killed by ghosts become ghosts", and nothing else, no vampire lord  "Drain of Life" ability.
It will always be possible to manage the power of ghosts later.
An example here : https://github.com/ihhub/fheroes2/pull/2003.
In summary:
"The suggestion can be summed up in two points:
- During combat, when a troop of ghosts obtains the hand, there is a 10% risk that it will attack one of the troops present in the hero's army.
- During combat, if the ghosts represent the greatest threat (the strongest troop) in the hero's army, there is a 10% risk that the ghosts will disappear from the battlefield."

**Edit 2023/03/16.**

Related discussion: https://github.com/ihhub/fheroes2/discussions/6849.

**Comment by ihhub** [comment]
Reclassifying this issue into a discussion based on the above conversation.

**Comment by LeHerosInconnu** [comment]
This is planned in my mod: https://github.com/ihhub/fheroes2/discussions/10007.
I close the discussion.
